packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ClientProvider.cs

```diff
@@ -31,18 +63,245 @@ public ClientProvider(InputClient inputClient)
+            ParameterProvider? endpointParam)
+        {
+            var isEndpointRequired = endpointParam != null && endpointParam.DefaultValue == null;
+            List<ParameterProvider> secondaryCtorParameters = isEndpointRequired ? [endpointParam, .. requiredParams] : requiredParams;
+
+            // initialize the arguments for the primary constructor
+            var primaryCtorArgs = new List<ValueExpression>(primaryCtorOrderedParams.Count);
+            foreach (var p in primaryCtorOrderedParams)
+            {
+                if (p.DefaultValue == null && p.Validation != ParameterValidationType.None)
+                {
+                    primaryCtorArgs.Add(p);
+                }
+                else
+                {
+                    primaryCtorArgs.Add(p.DefaultValue ?? DefaultOf(p.Type));
+                }
+            }
+
+            var primaryCtorInitializer = new ConstructorInitializer(false, primaryCtorArgs);
+            var constructorSignature = new ConstructorSignature(
+                Type,
+                _publicCtorDescription,
+                MethodSignatureModifiers.Public,
+                secondaryCtorParameters,
+                Initializer: primaryCtorInitializer);
+
+            return new ConstructorProvider(
+                constructorSignature,
+                MethodBodyStatement.Empty,
+                this);
+        }
+
+        private List<ParameterProvider> GetSubClientCtorParams()
+        {
+            var orderedParams = new List<ParameterProvider> { PipelineField.AsParameter };
+            ParameterProvider? endpointParam = null;
+
+            // add the auth parameter
+            if (_apiKeyAuthParameter != null)
+            {
+                orderedParams.Add(_apiKeyAuthParameter);
+            }
+
+            // add the client parameters, except the endpoint parameter
+            foreach (var parameter in _clientParameters)
+            {
+                if (parameter.Name == KnownParameters.Endpoint.Name)
+                {
+                    endpointParam = parameter;
+                }
+                else
+                {
+                    orderedParams.Add(parameter);
+                }
+            }
+
+            // order the constructor parameters as (Pipeline, Auth, ClientParams, Endpoint).
+            if (endpointParam != null)
+            {
+                orderedParams.Add(endpointParam);
+            }
+
+            return orderedParams;
+        }
+
+        private ConstructorProvider BuildMockingConstructor()
+        {
+            return new ConstructorProvider(
+                new ConstructorSignature(Type, $"Initializes a new instance of {Name} for mocking.", MethodSignatureModifiers.Protected, []),
+                new MethodBodyStatement[] { MethodBodyStatement.Empty },
+                this);
+        }
+
+        private ConstructorProvider BuildSubClientInternalConstructor()
```

Comment:
I would cut anything to do with sub clients and postpone for https://github.com/microsoft/typespec/issues/3690.

client ctors don't change due to client or sub client, the only thing that does change is if the ctors are internal which is a different property.  You can be a subclient with public ctors as well as a subclient with internal ctors.